### ISSUE TITLE
Proxy Use Case - Intercept API

### ISSUE THREAD

**Issue by jmrichardson** [trestletech/plumber#168]
Hi,

I would like to use plumber to "intercept" API calls to a service, transform some of the JSON, then forward to the service.  In other words, configure the client to point to plumber then any requests to a certain endpoint I would manipulate the request and send on to the external service.  All other endpoints would just be forwarded as is. 

Is this possible with plumber?  If so, any thoughts/example to implement this type of behavior?

Thanks in advance,
John

**Comment by trestletech** [comment]
It's technically possible, but you might be better off using a language/framework that natively supports HTTP proxying. Node, Go, Java, etc. all have packages that would allow you to just say: "proxy this incoming request to some external HTTP endpoint" and they have the intelligence to figure out what happens if the streams/connections are slow, if things get dropped, how to handle status codes and update the HTTP headers appropriately to mark it as a proxied request, etc.

So you could do it, but you'd be signing up for a big chunk of work.

**Comment by jmrichardson** [comment]
I agree that a proxy would be ideal.  But I am not aware of how to mutate/transform the message body in flight..  I went ahead and implemented it using plumber and it wasn't that bad..l.  I just created a filter.  Inside the filter I use R to change the postBody and then collect the client headers.  Then use httr to "proxy" the request to the service.  The httr response is sent back with the headers (using res$setHeader). I ran into an issue where plumber looks at the postBody, parses it if it's JSON and uses the result as args to the query string.  I had to fork plumber and change the parseBody function in post-Body.R to return an empty list (not parse into JSON).  That is, the postBody JSON is utilized by the service (not intended to be used for args).  Is there a way you can update plumber with a setting that disables this behavior?  

Thanks again for plumber!  So far it is working well and keeping up since there will only be 1 client per plumber pipeline.

**Comment by trestletech** [comment]
Nice! If you have that somewhere publicly available and didn't mind sharing a link I suspect others might be interested in seeing an implementation of a proxy in R. 

In terms of customizing the filter itself, I don't know that this would be a common enough use case to merit a top level config option on the plumber router itself, but if you're willing to dive into the "programmatic usage" (as opposed to the annotation-style most people use), you can customize which filters get installed, or provide your own filters. 

See https://www.rplumber.io/docs/programmatic-usage.html#creating-and-controlling-a-router , e.g.

> You can specify which filters you want in your new router by overriding the filters parameter when creating your new router.
